### PR TITLE
Introduce callables.

### PR DESCRIPTION
**Pull request by dan-zheng**
Introduce callables: values that can be "called" like functions.
Add a new `call` declaration kind to represent callable methods.

["Introducing callables" pitch](https://forums.swift.org/t/pitch-introduce-static-callables/21732). Proposal coming soon.

### CODE REVIEW

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -5885,7 +5903,53 @@ AbstractStorageDecl::AccessorRecord::getAccessor(AccessorKind kind) const {
...
+///
+/// \code
+/// struct Adder {
+///   var base: Int
+///   call(_ x: Int) -> Int {
+///      return base + x
+///   }
+/// }
+/// \endcode
+class CallDecl final : public FuncDecl {
```
Comment (dan-zheng) [inline include/swift/AST/Decl.h]: Conforming `CallDecl` to `FuncDecl` vs `AbstractFunctionDecl` is discussed in [this thread](https://forums.swift.org/t/question-what-should-calldecl-inherit-from/21815?u=dan-zheng).
It has a few nice properties - code sharing with `FuncDecl` is the primary one.

File: lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -3324,7 +3336,7 @@ parseIdentifierDeclName(Parser &P, Identifier &Result, SourceLoc &Loc,
                         StringRef DeclKindName, tok ResyncT1, tok ResyncT2,
                         tok ResyncT3, tok ResyncT4,
                         TokenProperty ResyncP1) {
-  if (P.Tok.is(tok::identifier)) {
+  if (P.Tok.is(tok::identifier) || P.Tok.is(tok::kw_call)) {
```
Comment (dan-zheng) [inline lib/Parse/ParseDecl.cpp]: Added to support parsing `call` as an identifier in non-declaration contexts.
This enables `func call` and `call(...)` (apply expressions).

File: lib/Sema/CSSimplify.cpp [hunk 3]
```diff
@@ -5228,6 +5228,107 @@ Type ConstraintSystem::simplifyAppliedOverloads(
...
+      return calculateForComponentTypes(protocolComp->getMembers());
+
+    // Otherwise, this must be a nominal type.
+    // Structural types cannot define `call` declarations.
+    auto nominal = canType->getAnyNominal();
+    if (!nominal) return llvm::DenseSet<CallDecl *>();
+
+    // If this type conforms to protocols, look up all of their `call`
+    // declaration requirements.
+    // FIXME: Fix this to handle overridden requirements, e.g. using
```
Comment (dan-zheng) [inline lib/Sema/CSSimplify.cpp]: TODO/WIP: This logic for "getting all call declarations" doesn't work properly.
It should mimic/be unified with `lookupQualified`, as mentioned in [this post](https://forums.swift.org/t/question-what-should-calldecl-inherit-from/21815/9?u=dan-zheng) by @slavapestov.

File: lib/Demangling/Demangler.cpp [hunk 4]
```diff
@@ -2829,6 +2829,8 @@ NodePointer Demangler::demangleFunctionEntity() {
       Args = TypeAndMaybePrivateName; Kind = Node::Kind::Allocator; break;
     case 'c':
       Args = TypeAndMaybePrivateName; Kind = Node::Kind::Constructor; break;
+    case 'F':
```
Comment (dan-zheng) [inline lib/Demangling/Demangler.cpp]: TODO: Add mangling/demangling tests.
Currently, there are no SIL tests for `call` declarations.
I informally checked that demangling works via `-emit-silgen`.
The old remangler might not work - haven't look into how to invoke/test it yet.

File: lib/Serialization/Serialization.cpp [hunk 5]
```diff
@@ -3409,6 +3410,69 @@ void Serializer::writeDecl(const Decl *D) {
...
+  // TODO: Revisit serialization.
+  // Create `CallLayout`? Not all fields from `FuncLayout` are necessary:
+  // `call` declarations cannot be static.
+  case DeclKind::Call: {
+    auto fn = cast<CallDecl>(D);
+    verifyAttrSerializable(fn);
+
+    auto contextID = addDeclContextRef(fn->getDeclContext());
+
+    unsigned abbrCode = DeclTypeAbbrCodes[FuncLayout::Code];
```
Comment (dan-zheng) [inline lib/Serialization/Serialization.cpp]: TODO: Create a `CallLayout` that is more specialized than `FuncLayout` (like `AccessorLayout`), dropping some fields like static spelling.

File: test/Syntax/Parser/tree.swift.result [hunk 6]
```diff
@@ -1,8 +1,8 @@
-<s118><s163><s92><s13><NULL/><NULL/><t6>// REQUIRES: syntax_parser_lib
+<s119><s164><s93><s13><NULL/><NULL/><t6>// REQUIRES: syntax_parser_lib
```
Comment (dan-zheng) [inline test/Syntax/Parser/tree.swift.result]: I’m not sure why changing this test to pass was necessary. I wonder if changing this test is breaking.

File: test/decl/call/protocol.swift [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  call() -> Self {
+    return self
+  }
+}
+struct S2 : P2 {}
+
+let s2 = S2()
+// TODO: Fix this by fixing `getCallDeclarations` logic in CSSimplify.cpp.
+// Need to handle overriding declarations; mimic/unify code with `lookupQualified`.
+// _ = s2()()()
```
Comment (dan-zheng) [inline test/decl/call/protocol.swift]: This is super critical to fix: relevant TODO comment in CSSimplify.cpp.

File: include/swift/IDE/SyntaxModel.h [hunk 8]
```diff
@@ -97,6 +97,7 @@ enum class SyntaxStructureKind : uint8_t {
   EnumElement,
   TypeAlias,
   Subscript,
+  Call,
```
Comment (johnfairh) [inline include/swift/IDE/SyntaxModel.h]: fyi will need a simple change to `ModelASTWalker::walkToDeclPre()` to actually emit this, and I think a possibly less-simple change to `SyntaxModelContext::SyntaxModelContext()` to classify `tok::kw_call` as identifer/keyword depending on context (this affects syntax highlighting).

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/Decl.h]: Committed to making `CallDecl` inherit from `FuncDecl`.

File: lib/Sema/CSSimplify.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/Sema/CSSimplify.cpp]: Done in https://github.com/apple/swift/pull/23517/commits/94a384feb522c132c6ec2e64f50daa53c940683a.

To find all `call` declarations, `TC.lookupMember` is used, which internally calls `lookupQualified`. This approach can be back-ported to `@dynamicCallable` and `@dynamicMemberLookup` to reduce code dupe (and improve correctness, in the case of `@dynamicCallable`). I'm interested in doing that in a follow-up.

File: test/decl/call/protocol.swift (hunk 7 above)
Comment (dan-zheng) [reply test/decl/call/protocol.swift]: Fixed in https://github.com/apple/swift/pull/23517/commits/94a384feb522c132c6ec2e64f50daa53c940683a.

File: lib/Serialization/Serialization.cpp (hunk 5 above)
Comment (dan-zheng) [reply lib/Serialization/Serialization.cpp]: Deferred for now. `CallDecl` has only two less fields than `FuncDecl` (static location and static spelling.)

**Comment by dan-zheng** [comment]
The main callable functionality is done and tested: ready for review.
Code completion and editor features require some more work.

**Comment by DougGregor** [comment]
@swift-ci build toolchain

**Comment by tkremenek** [comment]
@swift-ci please build toolchain

**Comment by dan-zheng** [comment]
@DougGregor @tkremenek 

Sorry, building failed because this PR depends on some minor changes to clang to support `call` declarations: https://github.com/apple/swift-clang/pull/284

I temporarily updated the clang version in `utils/update_checkout/update-checkout-config.json` to branch `callable` of `dan-zheng/swift-clang` so that building will work. If there's a better way, please let me know.

I've verified that all swift tests pass locally. Retriggering toolchain build.

**Comment by dan-zheng** [comment]
@swift-ci Please build toolchain

**Comment by dan-zheng** [comment]
@swift-ci Please build toolchain

**Comment by dan-zheng** [comment]
Investigating errors.

**Comment by dan-zheng** [comment]
@swift-ci Please build toolchain

**Comment by dan-zheng** [comment]
@swift-ci Please build toolchain

**Comment by dan-zheng** [comment]
@swift-ci Please build toolchain macOS

**Comment by dan-zheng** [comment]
@swift-ci Please build toolchain macOS

**Comment by dan-zheng** [comment]
@swift-ci Please benchmark

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
FWIW, SwiftLint (from the source compatibility suite) is failing with this error:

```
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/SwiftLint/Source/SwiftLintFramework/Rules/Style/MultilineFunctionChainsRule.swift:178:52: error: consecutive statements on a line must be separated by ';'
        return subcalls.flatMap { call -> [NSRange] in
                                                   ^
                                                   ;
```

**Comment by dan-zheng** [comment]
Closing, as [SE-0253](https://github.com/apple/swift-evolution/blob/master/proposals/0253-callable.md) was [returned for revision](https://forums.swift.org/t/returned-for-revision-se-0253-static-callables/23290).
Redesign using `func call` as the name for call-syntax delegate methods: https://github.com/apple/swift/pull/24299

**Pull request closed without merging**